modules/core/shared/src/main/scala/lucuma/core/model/sequence/gmos/binning.scala

```diff
@@ -0,0 +1,137 @@
+
+/**
+ * Optimal GMOS spectral and spatial binning calculations.
+ *
+ * See https://docs.google.com/document/d/1P8_pXLRVomUSvofyVkAniOyGROcAtiJ7EMYt9wWXB0o.
+ */
+object binning {
+
+  val DefaultSampling: PosDouble =
+    PosDouble.unsafeFrom(2.5)
+
+  /**
+   * Object angular size estimate based on source profile and image quality.
+   */
+  def objectSize(p: SourceProfile, iq: ImageQuality): Angle =
+    p match {
+      case SourceProfile.Point(_)          => iq.toAngle
+      case SourceProfile.Uniform(_)        => Angle.Angle180
+      case SourceProfile.Gaussian(fwhm, _) => fwhm
+    }
+
+  extension (λ: Angle) {
+    def arcsec: BigDecimal =
+      Angle.decimalArcseconds.get(λ)
+  }
+
+  /**
+   * Minimum of the slit and object size.
+   */
+  def effectiveWidth(
+    slitWidth:  Angle,
+    srcProfile: SourceProfile,
+    iq:         ImageQuality
+  ): Angle = {
+    given Order[Angle] = Angle.AngleOrder
+    slitWidth min objectSize(srcProfile, iq)
+  }
+
+  def spectralBinning(
+    slitWidth:  Angle,
+    srcProfile: SourceProfile,
+    iq:         ImageQuality,
+    dispersion: Quantity[Rational, NanometersPerPixel],
+    resolution: PosInt,
+    blaze:      Wavelength,
+    sampling:   PosDouble = DefaultSampling
+  ): GmosXBinning = {
+
+    // Effective resolution of slit
+    val effRes = resolution.value.toDouble / 2.0 / effectiveWidth(slitWidth, srcProfile, iq).arcsec
+
+    // Unbinned sampling
+    val nPix   = blaze.toNanometers.value.value/effRes/dispersion.value.toBigDecimal(4, HALF_UP)
+
+    // The maximum binning that gives the required sampling (<npix)
+    GmosXBinning.all.tail.reverse.find { bin =>
+      sampling.value < (nPix / bin.count)
+    }.getOrElse(GmosXBinning.One)
+  }
+
+  def northSpectralBinning(
+    fpu:        GmosNorthFpu,
+    srcProfile: SourceProfile,
+    iq:         ImageQuality,
+    grating:    GmosNorthGrating,
+    sampling:   PosDouble = DefaultSampling
+  ): GmosXBinning =
+    spectralBinning(fpu.effectiveSlitWidth, srcProfile, iq, grating.dispersion, grating.referenceResolution, grating.blazeWavelength, sampling)
+
+  def southSpectralBinning(
+    fpu:        GmosSouthFpu,
+    srcProfile: SourceProfile,
+    iq:         ImageQuality,
+    grating:    GmosSouthGrating,
+    sampling:   PosDouble = DefaultSampling
+  ): GmosXBinning =
+    spectralBinning(fpu.effectiveSlitWidth, srcProfile, iq, grating.dispersion, grating.referenceResolution, grating.blazeWavelength, sampling)
+
+  def spatialBinning(
+    srcProfile: SourceProfile,
+    iq:         ImageQuality,
+    pixelScale: Angle,
+    sampling:   PosDouble = DefaultSampling
+  ): GmosYBinning = {
+
+    // Number of unbinned pixels to sample the slit width
+    val npix = objectSize(srcProfile, iq).arcsec / (pixelScale.arcsec * sampling.value)
+
+    // The maximum binning that gives the required sampling (<npix)
+    GmosYBinning.all.tail.reverse.find { bin =>
+      bin.count <= npix
+    }.getOrElse(GmosYBinning.One)
```

Comment:
The python code ignores `bin - npix` values  under `-10.0`, which seemed rather arbitrary and I assumed it was just to make the maximum-finding search work out. Is there actually something significant about `-10.0` though? If so, `Uniform` source profile would probably result in y-binning of `One` instead of `Four`.